File: src/departed-db.ts
import { mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { homedir } from 'node:os';
import { join } from 'node:path';

export type DepartedRow = {
  userId: string;
  nickname: string;
  roomId: string;
  roomName: string;
  leftAt: string;
};

const ROOT = join(homedir(), '.loco-termux', 'rooms');

function roomDir(roomId: string): string {
  const safe = String(roomId).replace(/[^a-zA-Z0-9_-]/g, '_');
  return join(ROOT, safe);
}

function readRows(path: string): DepartedRow[] {
  try {
    const value = JSON.parse(readFileSync(path, 'utf8'));
    return Array.isArray(value) ? value : [];
  } catch (error) {
    const code = error && typeof error === 'object' && 'code' in error ? String((error as { code?: unknown }).code) : '';
    if (code !== 'ENOENT') console.error(`[DEPARTED-DB] ${path} 읽기 실패:`, error instanceof Error ? error.message : String(error));
    return [];
  }
}

function writeRows(path: string, rows: DepartedRow[]): void {
  mkdirSync(join(path, '..'), { recursive: true, mode: 0o700 });
  writeFileSync(path, JSON.stringify(rows, null, 2), { encoding: 'utf8', mode: 0o600 });
}

function paths(roomId: string): { current: string; history: string } {
  const dir = roomDir(roomId);
  return { current: join(dir, 'departed.json'), history: join(dir, 'leave-log.json') };
}

export function recordDepartures(roomId: string, roomName: string, ids: Array<string | number>, names: string[]): void {
  if (!ids.length) return;
  const { current, history } = paths(roomId);
  const currentRows = readRows(current);
  const historyRows = readRows(history);
  const now = new Date().toISOString();
  const added: DepartedRow[] = [];
  for (let i = 0; i < ids.length; i += 1) {
    const row: DepartedRow = {
      userId: String(ids[i]),
      nickname: String(names[i] || names[0] || '알 수 없음'),
      roomId: String(roomId),
      roomName,
      leftAt: now,
    };
    const index = currentRows.findIndex((item) => item.userId === row.userId);
    if (index >= 0) currentRows[index] = row;
    else currentRows.push(row);
    added.push(row);
  }
  writeRows(current, currentRows);
  writeRows(history, [...historyRows, ...added].slice(-10000));
}

export function listDeparted(roomId: string): DepartedRow[] {
  return readRows(paths(roomId).current).sort((a, b) => Date.parse(b.leftAt) - Date.parse(a.leftAt));
}

export function listLeaveHistory(roomId: string): DepartedRow[] {
  return readRows(paths(roomId).history).sort((a, b) => Date.parse(b.leftAt) - Date.parse(a.leftAt));
}

export function removeDeparted(roomId: string, userId: string): void {
  const path = paths(roomId).current;
  writeRows(path, readRows(path).filter((row) => row.userId !== String(userId)));
}

File: departed_db.py
from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path
import re

from storage import DATA_DIR, load_json, save_json, secure_dir

ROOMS_DIR = DATA_DIR / "rooms"
MAX_HISTORY = 10000


def _room_dir(room_id: str) -> Path:
    safe = re.sub(r"[^0-9A-Za-z._-]+", "_", str(room_id).strip())[:160] or "unknown"
    path = ROOMS_DIR / safe
    secure_dir(path)
    return path


def _now() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


def _read(path: Path) -> list[dict]:
    value = load_json(path, [])
    return value if isinstance(value, list) else []


def record_departure(room_id: str, room_name: str, user_id: str, nickname: str, message_id: str | None = None, count: int = 0) -> None:
    if not str(room_id).strip() or not str(user_id).strip():
        return
    directory = _room_dir(room_id)
    current_path = directory / "departed.json"
    history_path = directory / "leave-log.json"
    row = {"room_id": str(room_id), "room_name": str(room_name or ""), "user_id": str(user_id), "nickname": str(nickname or "알 수 없음"), "left_at": _now(), "message_id": message_id, "count": int(count or 0)}
    current = {str(x.get("user_id")): x for x in _read(current_path) if isinstance(x, dict) and x.get("user_id")}
    current[str(user_id)] = row
    history = _read(history_path)
    history.append(row)
    save_json(current_path, list(current.values()))
    save_json(history_path, history[-MAX_HISTORY:])


def list_departed(room_id: str) -> list[dict]:
    return sorted(_read(_room_dir(room_id) / "departed.json"), key=lambda x: str(x.get("left_at", "")), reverse=True)


def list_leave_history(room_id: str) -> list[dict]:
    return sorted(_read(_room_dir(room_id) / "leave-log.json"), key=lambda x: str(x.get("left_at", "")), reverse=True)


def remove_departed(room_id: str, user_id: str) -> None:
    path = _room_dir(room_id) / "departed.json"
    rows = [x for x in _read(path) if str(x.get("user_id", "")) != str(user_id)]
    save_json(path, rows)


def remove_all_departed(room_id: str, successful_ids: list[str]) -> None:
    if not successful_ids:
        return
    ids = {str(x) for x in successful_ids}
    path = _room_dir(room_id) / "departed.json"
    save_json(path, [x for x in _read(path) if str(x.get("user_id", "")) not in ids])
